Implementation and documentation of multi-threaded routing optimization

UAV-Pilot opened the pull request:
The implementation and usage are documented in README.md

andrasfuchs commented:
I'll need a few days to test the new parameters and check their results, but it looks good, excellent contribution, thank you @UAV-Pilot!

lynchaj commented:
Hi
Yes, this looks like an amazing new capability for FreeRouting. Using multiple cores for optimizing would be truly incredible!

Is there any chance to get a 32-bit Windows version so I can help with testing? I would love to see this running in person. I am still using 1.2.43 Extra with Java 8

Much appreciated, Thanks, Andrew Lynch
     On Tuesday, September 7, 2021, 06:59:43 AM EDT, Andras Fuchs ***@***.***> wrote:  
 
 


I'll need a few days to test the new parameters and check their results, but it looks good, excellent contribution, thank you @UAV-Pilot!


—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

andrasfuchs commented:
@lynchaj I'm not familiar with the build process here on GitHub, so I can't help you with the 32-bit question from the top of my head. We moved to Java 11 so you can't compile it for Java 8, but building a 32-bit version of the Java 11 executable should be possible, I think, but again I'm not an expert.

Could you try to clone @UAV-Pilot's [freerouting repo](https://github.com/UAV-Pilot/freerouting/tree/threaded_op), compile the source code and run it on your computer locally? It should work on 32-bit Windows too.

@miho, what do you think?
Since it was asked before, could we add 32-bit executables to our next release after this PR gets merged?

andrasfuchs reviewed (changes requested):
1, I run the same test input .DNS file with various -mt settings: 
a, without -mt (effectively using the default -mt 4), 
b, -mt 128 and
c, -mt 1. 
I got exactly the same .SES files with them and they all completed in 254-255 seconds. That is a little strange, isn't it? I got an 8-core CPU. 
The original v1.4.4 run in 612 seconds with the same input, and produced a slightly different .SES file.
Do you know what the reason can be?

2, Freerouter produced some temp files and they are not deleted after the successful run. Could you add a cleanup procedure?
![image](https://user-images.githubusercontent.com/910321/132381177-8b6e6930-e4ea-4de1-be6e-36e6d16398fb.png)

UAV-Pilot commented on src/main/java/eu/mihosoft/freerouting/logger/FRLogger.java:
```diff
@@ -10,7 +11,10 @@
 
 public class FRLogger {
-    private static Logger logger = LogManager.getLogger(FreeRouting.class);
+    private static Logger logger = LogManager.getLogger(FR.class);
```
Name "FreeRouting" alone is not too long, but "eu.mihosoft.freerouting.FreeRouting" was too long in log and made it harder to read log messages on screen. BTW, the original class FreeRouting is also an empty class. I actually have not worked with Java for quite long time except this project, so I'm not sure if it's easy to customize log4j for this behavior. If you're sure log4j can be customized easily for our purpose here, I'll revert it.

UAV-Pilot replied:
It's reverted anyway. Someone please take care of it later.

UAV-Pilot commented on src/main/java/eu/mihosoft/freerouting/gui/StartupOptions.java:
```diff
@@ -59,6 +65,29 @@ private void process(String[] p_args) {
                     }
+                } else if (p_args[i].startsWith("-mt")) {
+                    if (p_args.length > i + 1 && !p_args[i + 1].startsWith("-")) {
+                    	num_threads = Integer.decode(p_args[i + 1]);
```
will do

UAV-Pilot commented on src/main/java/eu/mihosoft/freerouting/gui/StartupOptions.java:
```diff
@@ -59,6 +65,29 @@ private void process(String[] p_args) {
+                } else if (p_args[i].startsWith("-us")) {
+                    if (p_args.length > i + 1 && !p_args[i + 1].startsWith("-")) {
+                    	String op = p_args[i + 1].toLowerCase().trim();
+                    	board_update_strategy = op.equals("global") ?
```
It's the same as  the existing ones

UAV-Pilot commented on README.md:
```diff
@@ -137,6 +137,10 @@ The following command line arguments are supported by freerouter:
 * -do [design output file]: saves a Specctra board (.dsn), a Specctra session file (.ses) or Eagle session script file (.scr) when the routing is finished
 * -mp [number of passes]: sets the upper limit of the number of passes that will be performed
 * -l [language]: "de" for German, otherwise it's English
+* -mt [number of threads]: sets thread pool size for route optimization
```
will do

UAV-Pilot commented:
With 8-cores, it's likely not meaningful to have more than 16 threads assuming each core is capable of hyperthreading. With such a simple board, the time for each thread to duplicate board is probably a large portion of execution time. Note that in README.md, I mentioned future enhancement about cloning.

It's better to use a more complex board to do the comparison test, e.g., a board previously needed more than 10 hours to optimize.

UAV-Pilot commented on src/main/java/eu/FR.java:
```diff
@@ -0,0 +1,9 @@
+package eu;
+
+public class FR {
```
deleted

lynchaj commented:
Hi

I’ve got a good example for you.  It’s my Z80 DUART board that’s been running for two days now.  I don’t know how many passes it has completed but it’s been well over a hundred.  2 layer PCB, all PTH construction.  Old school logic and passives all the way.  It solves on its own in the autorouter without any forcing or manual routing.  

 

It’s not done yet but getting there soon.  I sometimes run FreeRouting software early as a means to locate problems, to rearrange parts and move them closer together to improve overall trace routing.  Please give it a try and post the outcome.

 

See if you can beat 32 vias and 265.6795 inches of overall trace length

 

Thanks, Andrew Lynch

 

PS, I am running FR 1.2.43 Extra

 

From: UAV Pilot ***@***.*** 
Sent: Tuesday, September 7, 2021 3:12 PM
To: freerouting/freerouting ***@***.***>
Cc: lynchaj ***@***.***>; Mention ***@***.***>
Subject: Re: [freerouting/freerouting] Implementation and documentation of multi-threaded routing optimization (#64)

 

With 8-cores, it's likely not meaningful to have more than 16 threads assuming each core is capable of hyperthreading. With such a simple board, the time for each thread to duplicate board is probably a large portion of execution time. Note that in README.md, I mentioned future enhancement about cloning.

It's better to use a more complex board to do the comparison test, e.g., a board previously needed more than 10 hours to optimize.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/freerouting/freerouting/pull/64#issuecomment-914555399> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AUXGBYYTD4MFWCCQS26YI2DUAZPWPANCNFSM5DRIMDDQ> .
Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>  or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub> .   <https://github.com/notifications/beacon/AUXGBY247IQW6SDEZL7IYHLUAZPWPA5CNFSM5DRIMDD2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOG2BQEBY.gif>

lynchaj commented:
Argh! it stripped my attachment
     On Saturday, September 11, 2021, 11:31:00 AM EDT, lynchaj ***@***.***> wrote:  
 
 

Hi

I’ve got a good example for you. It’s my Z80 DUART board that’s been running for two days now. I don’t know how many passes it has completed but it’s been well over a hundred. 2 layer PCB, all PTH construction. Old school logic and passives all the way. It solves on its own in the autorouter without any forcing or manual routing. 

 

It’s not done yet but getting there soon. I sometimes run FreeRouting software early as a means to locate problems, to rearrange parts and move them closer together to improve overall trace routing. Please give it a try and post the outcome.

 

See if you can beat 32 vias and 265.6795 inches of overall trace length

 

Thanks, Andrew Lynch

 

PS, I am running FR 1.2.43 Extra

 

From: UAV Pilot ***@***.*** 
Sent: Tuesday, September 7, 2021 3:12 PM
To: freerouting/freerouting ***@***.***>
Cc: lynchaj ***@***.***>; Mention ***@***.***>
Subject: Re: [freerouting/freerouting] Implementation and documentation of multi-threaded routing optimization (#64)

 

With 8-cores, it's likely not meaningful to have more than 16 threads assuming each core is capable of hyperthreading. With such a simple board, the time for each thread to duplicate board is probably a large portion of execution time. Note that in README.md, I mentioned future enhancement about cloning.

It's better to use a more complex board to do the comparison test, e.g., a board previously needed more than 10 hours to optimize.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/freerouting/freerouting/pull/64#issuecomment-914555399> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AUXGBYYTD4MFWCCQS26YI2DUAZPWPANCNFSM5DRIMDDQ> .
Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub> . <https://github.com/notifications/beacon/AUXGBY247IQW6SDEZL7IYHLUAZPWPA5CNFSM5DRIMDD2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOG2BQEBY.gif> 



—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

UAV-Pilot commented:
An open source example board and documentation of approximate 27 times speedup with this board in README.md were checked in the last commit.

andrasfuchs commented:
@UAV-Pilot Thanks for the updates! Could you send me any of those files that had such big improvement in DSN formet, so that I can also test them?

The strange thing is so far that no matter how much threads I set to use (with the -mt) both the results and the time need are the same with my inputs. 

@lynchaj Could you somehow send me your Z80 DUART board as a DSN file? I would like to test it.
If GitHub prevents you from doing so, please try to zip it, or upload to a file sharing site, and include the link here. 
Thank you!

If you still want to join the testing, I would suggest you to use [IntelliJ IDEA](https://www.jetbrains.com/idea/) development environment, it works well for me. I'll continue running some test this week, so let me know if you need any help with the setup.

Also, since @miho seems to be not reacting to the 32-bit executable request, could you take a look and try to compile the source into a 32-bit version? I would start with the [gradlew.bat](https://github.com/freerouting/freerouting/blob/master/gradlew.bat), but I might be wrong.

UAV-Pilot commented:
@andrasfuchs The board file freerouting/tests/smoothieboard-v1.1_preOpt.bin was checked in, so it's available in my repository. As documented in README.md, this file was created in Freerouting v1.4.4, and bin format is native format of Freerouting, so Freerouting v1.4.4 should be able to open this file.

UAV-Pilot commented:
BTW, memory usage in route optimization of tests/smoothieboard-v1.1_preOpt.bin reached about 10 GB, so performance of route optimization for boards of medium to high complexity with 32-bit version will be affected by limited memory.

andrasfuchs commented:
Thanks for the file and the heads up about memory usage!

The bin file might be native, yes, but I would like to test it like it was in a normal scenario, passing it as a command line parameter, that's why I would prefer the board as DSN. Don't worry though, if you can't share it with me in that format, I will test the multi-threading functionality with other boards I have.

I'll try to make the time for it this week, and get back to you.

UAV-Pilot commented:
I tried to import Spectra Session file into KiCad and then to export DSN file, but an error prevented the import, so there is an bug somewhere in KiCad or freerouting with regarding importing Spectra Session file. While doing it, I noticed that previously I unintentionally left a ground zone on the top layer, so I went on to further change the imported KiCad project. Here are additional changes to the KiCad project: 1) removed a ground zone on the the top layer. 2) changes to resolve all DRC violations: changed footprints for several solder jumpers, shifted several R, C components, and moved J10, which seems to be in an odd place after the import. Before this pull request, I never used or worked on smoothieboard, so I don't know why some connectors were arranged in that way. In my latest commit, I put the whole KiCad project in zip format into the test folder. Inside it, there is an exported DSN file where majority of connections are not routed. Hopefully this zip file together with the DSN file inside can provide a starting point for anyone who likes to use it for testing. Actually any other publicly available board with enough complexity can also be used in test.

andrasfuchs commented:
I just run the following command:
`freerouting -mp 32 -de .\tests\smoothieboard-5driver.dsn -do .\tests\smoothieboard-5driver.ses -mt 32`

As I investigated the process it run only on one core (despite the -mt 32 parameter) and my CPU usage was only 10-12%.

Looking at the source code I would have expected to see the "Start to run optimization pass" in the logs, but it looks like the optimization didn't start.
@UAV-Pilot Do you have any good guess why? Limiting the number of passes (with -mp 32) could cause this?


I also got an error while I tried to import the SES file:
![image](https://user-images.githubusercontent.com/910321/133887858-2ef7a8f8-e3d8-48a1-a3d4-8559688613c3.png)
I can confirm that this error was not introduced by your pull request, it's a separate issue.

UAV-Pilot commented:
As mentioned in my previous comment, "there is an exported DSN file **_where majority of connections are not routed_**". You can also verify it via menu Info > Incompletes. 

It looks like importing from EAGLE to KiCad caused the SES importing issue. I did not encounter this problem with non-imported projects.

UAV-Pilot commented:
Revised the way to estimate speedup in the last commit. Previously time for Freerouting v1.4.4 to reach final via count and final trace length was used in the calculation.

UAV-Pilot commented:
DSN file within the zip was also updated

andrasfuchs commented:
Excellent, thank you. I'm playing around with the source code right now.

Could you get the latest commits into your pull request, so I can continue testing?
I found a bug that caused my previous confusion, but it looks like it's not in your pull request. It's related to the `-mp` parameter and I just want to make sure that your pull request will work as intended after it gets merged.

Thank you for your work and patience!

UAV-Pilot commented:
My latest commit is in the pull request. I just added a routed Smoothieboard (in bin format) which is ready for route optimization. It's smoothieboard-5driver_preOpt.bin in the zip file. After removing "    (component 
      (place @HOLE0 1854580 -557911 front 0)
      (place @HOLE1 863985 -557911 front 0)
      (place @HOLE2 928755 -1486280 front 0)
      (place @HOLE3 1853310 -1537080 front 0)
    )"  (Hopefully these are just mounting holes)
from the SES file, it can be imported although it might contain problems resulted from the above deletion. The resulted KiCad PCB board is: smoothieboard-5driver_preOpt.kicad_pcb in the zip file. As with any KiCad PCB files, routes exported to DSN are considered fixed by Freerouting by default, so it's a dead end to load DSN for route optimization unless you know a way in Freerouting to work around this restriction.

UAV-Pilot commented:
BTW, I never used the -mp option as all the boards except the smoothieboard that I used with Freerouting are boards designed by myself, and I wanted to get best possible PBC boards, so I did not put limit on passes.

andrasfuchs commented:
@UAV-Pilot I run some tests and your multi-threaded implementation works fine. I made a few small changes and fixed only a few small bugs. So I just want to thank you again for your contribution!

It would be great to see you contribute more, so don't hesitate to contact me if you feel like implementing another good idea of yours as a pull request.

UAV-Pilot commented:
@andrasfuchs I was kind of being forced to make this optimization when it seemed impossible to finish route optimization for one of my boards with previous version. Thanks for keeping Freerouting available and improving it!

AlexColello commented:
Out of curiosity, is there a paper describing the algorithm the multi-threaded optimizations are based on?

UAV-Pilot commented:
Currently all of document about the multi-threaded optimization are in README.md. Multi-threading is implemented via Java thread pool. Greedy board updating strategy is based on classic greedy algorithm. Prioritized item selection strategy is implemented via priority queue if I remember correctly. PCB routing optimization is complicated, and effectiveness of various implemented algorithms probably varies from board to board. If you're interested, please help improve existing ones and also add new ones.
